mitmproxy/websocket.py

```diff
@@ -1,168 +1,136 @@
 import warnings
 from typing import List
 from typing import Optional
-from typing import Union
 
-from mitmproxy import flow
+from mitmproxy import stateobject
 from mitmproxy.coretypes import serializable
-from mitmproxy.net import websocket
-from mitmproxy.utils import human
-from mitmproxy.utils import strutils
-
-from wsproto.frame_protocol import CloseReason
-from wsproto.frame_protocol import Opcode
 
 
 class WebSocketMessage(serializable.Serializable):
     """
-    A WebSocket message sent from one endpoint to the other.
+    A single WebSocket message sent from one peer to the other.
+
+    Fragmented WebSocket messages are reassembled by mitmproxy and the
+    represented as a single instance of this class.
+
+    The [WebSocket RFC](https://tools.ietf.org/html/rfc6455) specifies both
+    text and binary messages. To avoid a whole class of nasty type confusion bugs,
+    mitmproxy stores all message contents as binary. If you need text, you can decode the `content` property:
+
+    >>> if message.is_text:
+    >>>     text = message.content.decode()
     """
 
-    type: Opcode
-    """indicates either TEXT or BINARY (from wsproto.frame_protocol.Opcode)."""
     from_client: bool
     """True if this messages was sent by the client."""
-    content: Union[bytes, str]
+    is_text: bool
+    """
+    True if the message is a text message, False if the message is a binary message.
+
+    In either case, mitmproxy will store the message contents as *bytes*.
+    """
+    content: bytes
     """A byte-string representing the content of this message."""
     timestamp: float
     """Timestamp of when this message was received or created."""
 
-    killed: bool
-    """True if this messages was killed and should not be sent to the other endpoint."""
-
     def __init__(
         self,
-        type: int,
         from_client: bool,
-        content: Union[bytes, str],
+        is_text: bool,
+        content: bytes,
         timestamp: Optional[float] = None,
-        killed: bool = False
     ) -> None:
-        self.type = Opcode(type)  # type: ignore
         self.from_client = from_client
+        self.is_text = is_text
         self.content = content
         self.timestamp: float = timestamp or time.time()
-        self.killed = killed
 
     @classmethod
     def from_state(cls, state):
         return cls(*state)
 
     def get_state(self):
-        return int(self.type), self.from_client, self.content, self.timestamp, self.killed
+        return self.from_client, self.is_text, self.content, self.timestamp
 
     def set_state(self, state):
-        self.type, self.from_client, self.content, self.timestamp, self.killed = state
-        self.type = Opcode(self.type)  # replace enum with bare int
+        self.from_client, self.is_text, self.content, self.timestamp = state
 
     def __repr__(self):
-        if self.type == Opcode.TEXT:
-            return "text message: {}".format(repr(self.content))
+        if self.is_text:
```

Comment:
Currently in my add-on I'm explicitly checking both opcodes because the docs say:

> Currently there are two types of messages, corresponding to the BINARY and TEXT frame types.

and I didn't want to run in a future scenario were my `else` would catch others in a later version.

Does this change conclude that we'll never expose PING/PONG/CLOSE? I guess that's the only reason the opcode was exposed originally?

Reply:
Good point, we may want to leave that door open. I'll change it back to using the opcode. 👍